src/components/interface/cbufp/cbuf_c.c

```diff
@@ -0,0 +1,189 @@
+#include <cos_debug.h>
+#include <cos_alloc.h>
+#include <valloc.h>
+
+#define CSLAB_ALLOC(sz)   alloc_page()
+#define CSLAB_FREE(x, sz) free_page(x)
+#include <cslab.h>
+
+/* 
+ * All of the structures that must be compiled (only once) into each
+ * component that uses cbufs.
+ *
+ * I'd like to avoid having to make components initialize this data,
+ * so try and use structures that by default are initialized as all
+ * zero. 
+ */
+extern struct cos_component_information cos_comp_info;
+
+CVECT_CREATE_STATIC(meta_cbuf);
+PERCPU_VAR(cbuf_alloc_freelists);
+
+/*** Slow paths for each cbuf operation ***/
+
+/* 
+ * Precondition: cbuf lock is taken.
+ *
+ * A component has tried to map a cbuf_t to a buffer, but that cbuf
+ * isn't mapping into the component.  The component's cache of cbufs
+ * had a miss.
+ */
+int 
+__cbuf_2buf_miss(int cbid, int len)
+{
+	struct cbuf_meta *mc;
+	int ret;
+	void *h;
+
+	/* 
+	 * FIXME: This can lead to a DOS where the client passes all
+	 * possible cbids to be cbuf2bufed, which will allocate the
+	 * entire 4M of meta vectors in this component.  Yuck.
+	 * 
+	 * Solution: the security mechanisms of the kernel (preventing
+	 * cbids being passed to this component if they don't already
+	 * belong to the client) should fix this.
+	 */
+	mc = cbuf_vect_lookup_addr(cbid_to_meta_idx(cbid));
+	/* ...have to expand the cbuf_vect */
+	if (unlikely(!mc)) {
+		if (cbuf_vect_expand(&meta_cbuf, cbid_to_meta_idx(cbid))) BUG();    //???
+		mc = cbuf_vect_lookup_addr(cbid_to_meta_idx(cbid));
+		assert(mc);
+	}
+	ret = cbufp_retrieve(cos_spd_id(), cbid, len);
+	if (unlikely(ret < 0 || mc->sz < (len >> PAGE_ORDER))) return -1;
+	assert(CBUFM_GET_PTR(mc));
+
+	return 0;
+}
+
+static inline int
+__cbufp_alloc_slow(int cbid, int size, int *len, int *error)
+{
+	int amnt = 0, i;
+	static struct cbufp_shared_page *csp = NULL;
+
+	assert(cbid <= 0);
+	if (cbid == 0) {
+		struct cbuf_meta *cm;
+		struct cbufp_ring_element el;
+		if (!csp) csp = (struct cbufp_shared_page*)cbufp_map_collect(cos_spd_id());
+		/* Do a garbage collection */
+		amnt = cbufp_collect(cos_spd_id(), size);
+		if (amnt < 0) {
+			*error = 1;
+			return -1;
+		}
+		assert((unsigned)amnt <= CSP_BUFFER_SIZE);
+
+		if (amnt > 0) {
+			if (CK_RING_DEQUEUE_SPSC(cbufp_ring, &csp->ring, &el)) {
+				cbid = el.cbid;
+				/* own the cbuf we just collected */
+				cm = cbuf_vect_lookup_addr(cbid_to_meta_idx(cbid));
+				assert(cm);
+				assert(cm->cbid.cbid == cbid);
+				/* (should be atomic) */
+				CBUF_SET_TOUCHED(cm);
+				assert(CBUFM_GET_REFCNT(cm) < CBUF_REFCNT_MAX);
+				CBUFM_INC_REFCNT(cm);
+			} else {
+				/* Someone stole the cbufs I collected! */
+				amnt = 0;
+			}
+		}
+		/* ...add the rest back into freelists. construct a temporal freelist 
+		 * first, then add this temporal list to freelist atomically*/
+		struct cbuf_meta *meta, *tail, *head, old_head, new_head;
+		int cb, idx;
+		if (CK_RING_DEQUEUE_SPSC(cbufp_ring, &csp->ring, &el)) {
```

Comment:
This is all rather dense code.  Can you add in a few empty lines to separate out the separate thoughts?